// @sitecore-feaas/clientside/src/ui/FEAASPicker.tsx
import type * as FEAASPickerImports from '@sitecore-feaas/frontend/picker'
import { FEAASBase } from '../components/FEAASBase.js'

export function FEAASPickerProps(element: FEAASBase<any, any> = this) {
  return {
    part: element?.getAttribute('part') || 'picker',
    slot: element?.getAttribute('part') || 'picker',
    mode: element?.getAttribute('mode') as 'library' | 'element' | 'themes',
    placeholder: element?.getAttribute('placeholder') || 'No theme',
    frontend: element?.getContextAttribute('frontend') ?? 'https://components.sitecorecloud.io',
    src: element?.getAttribute('src') ?? '/picker.js',
    hidden: element?.getAttribute('hidden') != null,
    variant: element?.getAttribute('variant') as 'internal' | 'pages',
    context: undefined as FEAASPickerImports.PickerProps['context'],
    onClose: undefined as (...args: any[]) => void,
    onContextAttributeSet: undefined as (context: any, attribute: string, value: string) => void
  }
}
export type FEAASPickerProps = ReturnType<typeof FEAASPickerProps>

export class FEAASPicker extends FEAASBase<FEAASPickerProps, typeof FEAASPickerImports> {
  static observedAttributes = ['library', 'mode', 'frontend', 'src', 'hidden', 'variant']

  defaultProps = {
    hidden: false
  } as Partial<FEAASPickerProps>

  getProps() {
    return FEAASPickerProps(this)
  }

  load() {
    return this.import(this.formatURL(this.props.src, this.props.frontend))
  }

  render(
    { placeholder, context, mode, children, variant }: React.PropsWithChildren<FEAASPickerProps>,
    { Picker, React, ReactDOM }: typeof FEAASPickerImports
  ) {
    return (
      <Picker
        placeholder={placeholder}
        mode={mode || undefined}
        onClose={this.props.onClose}
        onContextAttributeSet={this.props.onContextAttributeSet}
        context={context}
        variant={variant}
      >
        {children}
      </Picker>
    )
  }
}
declare global {
  namespace JSX {
    interface IntrinsicElements {
      'feaas-picker': Partial<typeof FEAASPickerImports>
    }
  }
}

FEAASPicker.register('feaas-picker')
