// @sitecore-feaas/clientside/build/ui/FEAASPicker.js
import { FEAASBase } from '../components/FEAASBase.js';
export function FEAASPickerProps(element = this) {
    var _a, _b;
    return {
        part: (element === null || element === void 0 ? void 0 : element.getAttribute('part')) || 'picker',
        slot: (element === null || element === void 0 ? void 0 : element.getAttribute('part')) || 'picker',
        mode: element === null || element === void 0 ? void 0 : element.getAttribute('mode'),
        placeholder: (element === null || element === void 0 ? void 0 : element.getAttribute('placeholder')) || 'No theme',
        frontend: (_a = element === null || element === void 0 ? void 0 : element.getContextAttribute('frontend')) !== null && _a !== void 0 ? _a : 'https://components.sitecorecloud.io',
        src: (_b = element === null || element === void 0 ? void 0 : element.getAttribute('src')) !== null && _b !== void 0 ? _b : '/picker.js',
        hidden: (element === null || element === void 0 ? void 0 : element.getAttribute('hidden')) != null,
        variant: element === null || element === void 0 ? void 0 : element.getAttribute('variant'),
        context: undefined,
        onClose: undefined,
        onContextAttributeSet: undefined
    };
}
export class FEAASPicker extends FEAASBase {
    constructor() {
        super(...arguments);
        this.defaultProps = {
            hidden: false
        };
    }
    getProps() {
        return FEAASPickerProps(this);
    }
    load() {
        return this.import(this.formatURL(this.props.src, this.props.frontend));
    }
    render({ placeholder, context, mode, children, variant }, { Picker, React, ReactDOM }) {
        return (React.createElement(Picker, { placeholder: placeholder, mode: mode || undefined, onClose: this.props.onClose, onContextAttributeSet: this.props.onContextAttributeSet, context: context, variant: variant }, children));
    }
}
FEAASPicker.observedAttributes = ['library', 'mode', 'frontend', 'src', 'hidden', 'variant'];
FEAASPicker.register('feaas-picker');
